MAINT, ENH: Refactor percentile and quantile methods

abel-bzz opened the pull request:
This is quite a big pull request especially for a new comer like me. 
I hope it respect the standards of numpy.

This PR should solve #10736.
It is a rewrite of the work done on https://github.com/Ouranosinc/xclim/pull/802, with some other cases handled.

## API changes
The interpolation option for quantile/percentile has been expanded.
Interpolation can be passed as a string or a value of QuantileInterpolation Enum.
This gives more flexibility to add new interpolation methods and is clearer than using magic strings.

Note: In the issue #10736, it was asked to remove some existing methods but I don't think it should be part of the same PR. 

## Algorithm changes
### indices computation
The logic has changed a bit for to compute `indices` (a.k.a `virtual_indices`).
Instead of doing `q * (Nx - 1)`, it is now `(q * Nx) - 1` which seems more reasonable, see below example.
However, to not break the existing behavior of nearest, midpoint, lower and higher, I kept it as is, even if it is probably giving wrong results.
<details>
  <summary>Example </summary>

  Let's say you want the *50th* percentile of a 100 size ordered list.  
  If the array index start at zero,
  doing `0.5 * ( 100 -1 )` would give `49.5`
  which is wrong because, the 50th percentile is *exactly* on the 49th index.
  But doing `0.5 * 100 - 1` gives the expected `49`
</details>

**This has no consequence on the result of the default interpolation method (method 7). This is because it is taken into account with the formula using alpha == beta == 1**

### Rewrite of existing interpolation methods 
The existing nearest, midpoint, lower and higher methods have been transformed to fit the QuantileInterpolation model.


## Performance improvements
A performance improvement could be done for nanpercentile but it's no longer part of this PR.
~The performance of nanpercentile/nanquantile have been significantly improved, especially when the wanted `quantiles` is a scalar or a short list of quantiles. In most case it should perform around **50 times** faster than before.
I will add here a link to the performance report once I re-run the perf tests, in the meantime you can find the report of the Xclim version here: https://gist.github.com/bzah/2a84d050b8a1aed1b40a2ed1526e1f12. It might not be accurate thought.~

<!--         ----------------------------------------------------------------
                MAKE SURE YOUR PR GETS THE ATTENTION IT DESERVES!
                ----------------------------------------------------------------

*  FORMAT IT RIGHT:
      http://www.numpy.org/devdocs/dev/development_workflow.html#writing-the-commit-message

*  IF IT'S A NEW FEATURE OR API CHANGE, TEST THE WATERS:
      http://www.numpy.org/devdocs/dev/development_workflow.html#get-the-mailing-list-s-opinion

*  HIT ALL THE GUIDELINES:
      https://numpy.org/devdocs/dev/index.html#guidelines

*  WHAT TO DO IF WE HAVEN'T GOTTEN BACK TO YOU:
      http://www.numpy.org/devdocs/dev/development_workflow.html#getting-your-pr-reviewed
-->

abel-bzz committed:
MAINT, ENH [#10736] Add interpolation methods to quantile

- Added the missing linear interpolation methods.
- Updated the existing unit tests.

- Added pytest.mark.xfail for boolean arrays
See
- https://github.com/numpy/numpy/pull/19857#issuecomment-919258693
- https://github.com/numpy/numpy/issues/19154

bzah commented on numpy/lib/tests/test_perf.py:
```diff
@@ -0,0 +1,29 @@
+import unittest
```
Oops, this file is not supposed to be commited

BvB93 commented:
Just as an initial comment: this PR contains _a lot_ of stylistic changes to the existing code base that are unrelated to the `np.quantile` reword. These changes shouldn't really be in this PR, and, honestly, the sheer amount of them makes the reviewing process a lot more difficult.

seberg commented:
Nice!  Sounds like there are a couple of nuts to crack still.  Some notes:

* I do think we have to retain the dtype for all non-interpolating methods (R may not care about this) as you noted.
* The `fuzz` thing has me slightly worried, but I need to have a closer look.  It feels like this may depend on the data type of `q` passed in by the user which is not ideal?
* About the stacklevel:  Generally it is 2 + each level of additional helper function.  In this case the `@array_function` decorator adds one level, so its 3+each level.  4 is probably right.  You can test this manually: If you trigger it in a session, it should point to your line of code, if you lower it by one, it should then point to somewhere inside NumPy.
* I think you added `drop_nans` for the internal API?  Could you "drop" that ;)? I don't doubt that it is useful to speed up `nanquantile`, but it adds complexity to an already complicated change.

Some API things we probably need to hash out, but I think these can be discussed later:
* My instinct is to add a new `method=` keyword argument and deprecate `interpolation` entirely.  This probably means adding `method="legacy_nearest"` or so that code can retain the (discouraged) behaviour if necessary and avoid the warning.  (But this is a discussion for the mailing list and I think the current design makes it nice to modify it.)
* What happens if the percentile is outside of the data range?  My first instinct would be to raise an error, because it is easier to go from error -> extrapolation than the other way around.  I think this may be your last point?  I think some of the non-interpolating methods must fail for `quantile={0, 1}`.

bzah commented:
> 
> 
> Just as an initial comment: this PR contains _a lot_ of stylistic changes to the existing code base that are unrelated to the `np.quantile` reword. These changes shouldn't really be in this PR, and, honestly, the sheer amount of them makes the reviewing process a lot more difficult.

Yes you are right, I'm used to run a auto linter all the time but I understand that it make the review difficult.
I will revert the unrelated changes and take care of the comments of @seberg tomorrow.

About `drop_nans` I completely agreed that it make the code barely readable. At least for the initial review I will remove it and maybe later if the performance boost is deemed worth it could be re-added or refactored in.

bashtage commented on numpy/lib/function_base.py:
```diff
@@ -4014,122 +4116,233 @@ def _quantile_is_valid(q):
+   Note:
+       Compared to R, -1 is added because R array start at 1 (0 for python)
+   """
+    return array_size * quantiles + (
```
Have you looked at R source?  If so, this could be a problem since it is GPL, and so incompatible with NumPy's license.

bzah replied:
I've looked into several source codes including the one of R. I thought the comment would make sense because R source code for quantile was the reference (at least for me). But I can totally drop the comment it if it may cause license issues.
It doesn't add much to the comprehension anyway.

bashtage replied:
The comment isn't the problem.  Any values lifted from source code is the problem.  In principle, anything that has been directly influenced by R source code should be removed from the PR, and a clean sheet implementation of these features would be needed to steer clear of GPL risk.  Anything the can be verified using documentation alone, or by experimenting with R software, would be OK (IANAL of course)

Ultimately I think the correct approach is really a question for @charris, @rgommers and other core maintainers about how this type of issue should be handled.

bzah replied:
On this specific case, the formula can be sourced with , for example, Hyndman&Fan paper "Sample Quantiles in Statistical Packages" in The American Statistician, Vol. 50, No. 4 (Nov., 1996), pp. 361-365
It is specific to quantile interpolations regardless of the language.

eric-wieser replied:
I suspect this comment is here because the comment specifically mentions R, and is not actually about `virtual_index_formula`

bashtage replied:
@bzah That is good news.  If your entire formula can be reproduced from the paper, then you can just cite the paper as the source and remove any other citations.

seberg replied:
If this may be 100% clean license wise, lets consider pinging the original R authors to give their OK (the R code and that paper are the authors).  I will assume the R code did not changed in relevant ways since then.

bashtage commented on numpy/lib/function_base.py:
```diff
@@ -4014,122 +4116,233 @@ def _quantile_is_valid(q):
+    # fuzz avoid edge cases where a upper bound would be wrongly chosen due
+    #  to a failed floating point comparison
+    # TODO find out why epsilon * 4 instead of just epsilon
+    #  (it is done like this in R and in climdex c++ impl)
```
It looks here as well that you are specifically referring the R source code.

bzah replied:
This fuzz part is subject to discussion. I did an update on this in 47e8a7d4012b6c53fbae318c7ccf46ed364654aa but if necessary I guess it can be removed even if it may fix some edge cases.
This correction was not done in the former version of quantile so it could stay that way.

bzah replied:
Fuzz was removed, I'm resolving this.

eric-wieser commented on numpy/lib/function_base.py:
```diff
@@ -4014,122 +4116,232 @@ def _quantile_is_valid(q):
+    ) - 1
+
+
+def gamma_formula(val: np.array, val_floor: np.array) -> np.array:
+    """
+    Compute the gamma (a.k.a 'm' or weight) for the linear interpolation
+    of quantiles.
+    For the extreme values 0 and 1,
+     when gamma is 0 the left bound will be picked,
+     when gamma is 1 the right bound will be.
+    """
+    return val - val_floor
```
This is a bizarre way to spell subtraction. Why not just use subtraction?

bzah replied:
In the most simple cases, gamma is just x - floor(x) but depending on the interpolation method it may be different.
I will refactor this into to have a `::get_gamma` which will handle all the gamma cases in one place.

bzah commented on numpy/lib/tests/test_function_base.py:
```diff
@@ -2817,18 +2817,91 @@ def test_2D(self):
+        assert np.all(res == 27)
+
+    def test_linear_normal_unbiased(self):
+        # METHOD 8 of H&F
```
copy pasta typo, should be METHOD 9

BvB93 commented on numpy/lib/function_base.py:
```diff
@@ -4014,122 +4135,234 @@ def _quantile_is_valid(q):
+    # makes them problematic for __setitem__ and attribute access. As a
+    # workaround, we call this on the result of every ufunc on a possibly-0d
+    # array.
+    not_scalar = np.asanyarray
```
I'd suggest removing this alias. I know it was in the original code, but I feel it really doesn't add anything worthwhile.

BvB93 commented on numpy/lib/function_base.py:
```diff
@@ -4014,122 +4135,234 @@ def _quantile_is_valid(q):
-        r = take(ap, indices, axis=0, out=out)
-
+    arr = np.moveaxis(arr, axis, destination=0)
+    valid_values_count = np.broadcast_to(not_scalar(values_count),
+                                         arr.shape[1:])
```
```suggestion
    valid_values_count = np.broadcast_to(values_count, arr.shape[1:], 
                                         subok=True)
```
No need to explicitly call `asanyarray` here, just let `broadcast_to` do it under the hood.

BvB93 commented on numpy/lib/tests/test_function_base.py:
```diff
@@ -2817,18 +2817,91 @@ def test_2D(self):
 
-    def test_linear(self):
-
+    def test_linear_inverted_cdf(self):
+        # METHOD 1 of H&F
+        arr = np.asarray([15.0, 20.0, 35.0, 40.0, 50.0])
```
```suggestion
    TYPE_CODES = np.typecodes["AllInteger"] + np.typecodes["AllFloat"] + "O"

    @pytest.mark.parametrize("dtype", TYPE_CODES)
    def test_linear_inverted_cdf(self, dtype):
        # METHOD 1 of H&F
        arr = np.asarray([15.0, 20.0, 35.0, 40.0, 50.0], dtype=dtype)
```
It's generally a good idea to check with multiple input dtypes.

BvB93 commented on numpy/lib/function_base.py:
```diff
@@ -4014,122 +4135,234 @@ def _quantile_is_valid(q):
+        if np.issubdtype(arr.dtype, np.inexact) and np.isnan(result):
+            warnings.warn("All-NaN slice encountered", RuntimeWarning,
+                          stacklevel=4)
+        return np.broadcast_to(result, (quantiles.size,) + result.shape)
```
```
 >           return np.broadcast_to(result, (quantiles.size,) + result.shape)
E           AttributeError: 'int' object has no attribute 'shape'
```
This can fail if `arr` is an object array, as `result` is thus not guaranteed to have the `shape` attribute.

bzah replied:
Good catch, I thought arr was wrapped in `asarray` above, but I might have lost it at some point.

BvB93 commented on numpy/lib/function_base.py:
```diff
@@ -4014,122 +4135,234 @@ def _quantile_is_valid(q):
+        interpolation.virtual_index_fun(valid_values_count, quantiles),
+    )
+    # After sort, slices having NaNs will have for last element a NaN
+    previous_indexes = np.floor(virtual_indexes)
+    next_indexes = previous_indexes + 1
+    # When indexes is above max index, take the max value of the array
+    indexes_above_bounds = virtual_indexes >= valid_values_count - 1
+    if indexes_above_bounds.any():
+        previous_indexes[indexes_above_bounds] = -1
+        next_indexes[indexes_above_bounds] = -1
+    # When indexes is below min index, take the min value of the array
+    indexes_below_bounds = virtual_indexes < 0
+    if indexes_below_bounds.any():
+        previous_indexes[indexes_below_bounds] = 0
+        next_indexes[indexes_below_bounds] = 0
+    if np.issubdtype(arr.dtype, np.inexact):
+        virtual_indexes_nans = np.isnan(virtual_indexes)
+        if virtual_indexes_nans.any():
+            previous_indexes[virtual_indexes_nans] = -1
+            next_indexes[virtual_indexes_nans] = -1
+    previous_indexes = previous_indexes.astype(int)
+    next_indexes = next_indexes.astype(int)
```
Speaking of splitting things into smaller functions: the generation of `next_indexes` and `previous_indexes` sounds like a prime candidate for their own dedicated function.

bzah replied:
Done

bzah commented on numpy/lib/function_base.py:
```diff
@@ -52,6 +53,107 @@
     ]
 
 
+class QuantileInterpolation(Enum):
```
I think it would be helpful if it remains public.
If someone want to extend the enum with it's own alpha and beta because none of the existing method fit his/her needs, having the enum public would allow that.

BvB93 replied:
Personally I'd be okay with its addition.
Nevertheless, this could definetely use some further discussion, especially on the namespaces where we'd want to expose it (`np`, `np.lib`, `np.lib.function_base`?).

seberg replied:
We discussed this a bit in the triage call, and I think the gist of it was: Do we really want to have both strings and the enum?  And if we have to pick, at least I still tend towards preferring strings.  (Which right now is also the backcompatible method).

(We did not have a very strong opinion on this though)

About using `interpolation=` or `method=`, I still think I prefer method. But there is a good point that it makes a good followup.  I have to take some quality time on the rest of the PR, before I want to do a final bike-shedding on the names.

BvB93 replied:
```suggestion
class _QuantileInterpolation(Enum):
```
Continuing from https://github.com/numpy/numpy/pull/19857#discussion_r714214190:
It would be best for now to make the enum private (and remove all mention of it from the release notes & public API docs).

bzah commented on numpy/lib/function_base.py:
```diff
@@ -52,6 +53,107 @@
+            virtual_index_fun,
+            get_gamma_correction,
+    ):
+        self.method_name = aliases
```
I forgot to renamed it but here, but I use "aliases" for the constructor parameter, would it be ok ?

bzah commented:
I did quite a big refactoring, the implementation is now much closer to what the former *_quantile* was and is hopefully much easier to understand.
I still have a single unit test to fix before moving from draft to publish.

bzah commented:
@seberg What is the expected behavior of `_quantile` when the input array is made of booleans ?
I'm asking this, because the failing unit test is a specific case of #19154. 

The unit test was previously handled in a strange manner using `if np.issubdtype(indices.dtype, np.integer)`. 
Having indices as integers will almost never happen in real life IMO, especially when alpha and beta are floats.

So if I understand what is the expected behavior for booleans I could try to fix the unit test and #19154 in one commit.
Otherwise I can re-add `if np.issubdtype(indices.dtype, np.integer)` to workaround the unit test.

bashtage commented:
@bzah Seems that `quantile` doesn't work at all for boolean arrays, although I don't think this is probably the correct behavior.

```
np.quantile(np.array([1,0,0,0,1,0,0,0,1,0,0,0],dtype=bool),0.25)
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
...
TypeError: numpy boolean subtract, the `-` operator, is not supported, use the bitwise_xor, the `^` operator, or the logical_xor function instead.
```

bzah commented:
> @bzah Seems that `quantile` doesn't work at all for boolean arrays, although I don't think this is probably the correct behavior.
> 
> ```
> np.quantile(np.array([1,0,0,0,1,0,0,0,1,0,0,0],dtype=bool),0.25)
> ---------------------------------------------------------------------------
> TypeError                                 Traceback (most recent call last)
> ...
> TypeError: numpy boolean subtract, the `-` operator, is not supported, use the bitwise_xor, the `^` operator, or the logical_xor function instead.
> ```

Yes, in the former implementation of _quantile it only works with booleans if the indices are integers. 
In other words it works if `(ap.shape[axis] * q ) % 1 == 0`
I think the only unit test for booleans is *numpy/lib/tests/test_function_base.py::TestQuantile::test_correct_quantile_value* 
and it is a specific case of this.

seberg commented:
There was a change, I think it used to work, but probably we can get away with doing what we think is right.  My feeling is that the correct thing is to return the input dtype unmodified for any of the non-interpolating methods.  For the interpolating methods, you will get whatever `bool * float` which is a float.

bzah commented:
Well I don't get it.
Given:
```
arr = np.asarray([True,False])
result = np.percentile(arr, 0.5, interpolation="any linear interpolation")
```
If arr is cast into integers, this would give a result close to 0.5.
What can the user make of it ?
Does it really makes sense to have a float result for booleans input ?

I guess if someone has a real world use case for percentile on booleans it would help me understand it.

seberg commented:
> Does it really makes sense to have a float result for booleans input ?

No, to be honest, it doesn't.  But NumPy (similar to Python) usually pretends that booleans are like integers and thus "numerical".  Since it is already an error right now, I don't mind if you keep it there until a clear decision is made though.

So the argument is that `True * 0.5` or `np.mean([True, True])` make just as little sense, but both mostly work.  We just apply the same (currently typical) logic to `median`.  From a type safety perspective boolean probably should not be numerical, this is inherited from Python (which may well have got it because it did not have booleans in Python 1, I believe!)

bzah commented:
Ok ok thanks @seberg ! 
I'm adding `@pytest.mark.xfail` to the broken unit test with a ref to #19154 
I think the PR will be ready to leave draft then.

Thanks a lot for all the reviews already done !

bzah commented:
I'm removing the `fuzz` correction.
When fuzz == epsilon * 4 (what is done in R), it is only to handle some edge cases where virtual_index is < 8.
When  fuzz == epsilon (what I did), it handles virtual_index is < 2 which is worthless.

Moreover, by "handling" I mean it will force the result to be the left bound instead of interpolating to something very close to the left bound.
So unless I'm missing something, I really don't think it is necessary.

This also means there should not be any other issue with the license as far as I understand it.

bzah commented:
On the CI, I get a cygwin error:
```
 Maximum tries of 3 reached. Throwing error.
Could not move the bad package to the failure directory. It will show as installed.
 C:\ProgramData\chocolatey\lib\Cygwin
 The error:
 (32) The process cannot access the file because it is being used by another process: [\\?\C:\ProgramData\chocolatey\lib-bad\Cygwin\tools\setup-x86_64.exe]
```
 Is this usual ? Can I do something to fix it ?

charris commented:
You can ignore the Cygwin error, it is on the CI side.

charris commented:
Could you add a release note? It would go in `doc/release/upcoming_changes` where you can also see examples.

BvB93 commented on numpy/lib/function_base.py:
```diff
@@ -3,6 +3,8 @@
 import sys
 import warnings
+from enum import Enum
+from typing import Tuple
```
```suggestion
```
If you're interested: annotations (of the public API) should go in the [function_base.pyi](https://github.com/numpy/numpy/blob/main/numpy/lib/function_base.pyi) stub file.

BvB93 commented on numpy/lib/function_base.py:
```diff
@@ -4014,122 +4193,207 @@ def _quantile_is_valid(q):
+        if virtual_indexes_nans.any():
+            previous_indexes[virtual_indexes_nans] = -1
+            next_indexes[virtual_indexes_nans] = -1
+    previous_indexes = previous_indexes.astype(int)
+    next_indexes = next_indexes.astype(int)
```
```suggestion
    previous_indexes = previous_indexes.astype(np.intp)
    next_indexes = next_indexes.astype(np.intp)
```
As a general rule of thumb: use `intp` if you're creating an array of indices. The problem with `int` is that it can map to either `int64` or `int32` depending on the platform (say hi Windows), the latter of which is more prone to overflow when indexing an array.

BvB93 commented on numpy/lib/tests/test_function_base.py:
```diff
@@ -2817,18 +2817,94 @@ def test_2D(self):
-
+    TYPE_CODES = np.typecodes["AllInteger"] + np.typecodes["AllFloat"] + "O"
+
+    @pytest.mark.parametrize("dtype", TYPE_CODES)
+    def test_linear_inverted_cdf(self, dtype):
+        # METHOD 1 of H&F
+        arr = np.asarray([15.0, 20.0, 35.0, 40.0, 50.0], dtype=dtype)
```
Could you apply this parametrization to all the newly added tests?

bzah replied:
oops I forgot that

bzah replied:
For information, when dtype is complex256 or float128 the result is not always exactly equal to the expected result.
For example with on `test_linear_median_unbiased` it gives `27.000000000000001776+0j)` instead of `27`.

Thus I changed the assertions to` np.testing.almost_equal` with either 14 or 15 precision to workaround this.

eric-wieser commented on numpy/lib/function_base.py:
```diff
@@ -4014,122 +4192,212 @@ def _quantile_is_valid(q):
+        return unknown
+    elif isinstance(unknown, str):
+        for inter in QuantileInterpolation:
+            if unknown.lower() in inter.aliases:
```
Are you sure we should be calling `lower` here? Did we support mixed case before?

seberg commented on numpy/lib/function_base.py:
```diff
@@ -4014,122 +4192,212 @@ def _quantile_is_valid(q):
+        if result.ndim == 0 and out is None:
             # can't write to a scalar
-            r = a.dtype.type(np.nan)
+            result = np.asanyarray(np.nan)
```
```suggestion
            result = np.array(np.nan, dtype=result.dtype)
```
I guess this needs some other type, such as `a.dtype` here.  Returning `np.nan` seems incorrect, since it will always be double precision.  `result.dtype` may not work for `object` dtype, but I guess that case is wrong anyway, since it does not handle NaNs anyways (also before these changes).

seberg replied:
A test using `float128` as input may be good.

bzah replied:
I added a unit test for this using `AllFloat` and your suggestion works better than what I wrote but I still have a few cases going wrong.
It goes:
from `float16` to `float64`; 
from `float32` to `float64`
and from `float64` to `float128`

I will try to fix this as well.

bzah replied:
I fixed it using `arr.dtype` instead of `result.dtype`. It should be safe thanks to `if np.any(slices_having_nans):` above.

charris commented:
Anymore comments before this goes in?

seberg commented on numpy/lib/function_base.py:
```diff
@@ -4014,122 +4192,212 @@ def _quantile_is_valid(q):
+    These methods are extended to this function using _ureduce
+    See nanpercentile for parameter usage
+    It computes the quantiles of the array for the given axis.
+    A linear interpolation is performed based on the interpolation_type.
```
```suggestion
    A linear interpolation is performed based on the `interpolation`.
```

seberg commented on numpy/lib/function_base.py:
```diff
@@ -4014,122 +4192,212 @@ def _quantile_is_valid(q):
+                       interpolation)
+    result_shape = virtual_indexes.shape + (1,) * (arr.ndim - 1)
+    gamma = gamma.reshape(result_shape)
+    result = _linear_interpolation_formula(previous,
```
The old code does not call the interpolation formula when the dtype of the indices is integer I think (i.e. gamma is unimportant?)

In particular:
```
np.quantile(np.array([1, 2], dtype=np.uint8), [0.5], interpolation="nearest")
```
still returns integers (because you do cast to `intp` in the `gamma` formula) but not `uint8` anymore.  (Or am I missing the bool regression thing in current NumPy that was rolled back in main, but not here?)

bzah replied:
Yes previously when dtype of indices was integer it was short-cutting the interpolation because with an integer we can directly pickup the element in the array whichever is the interpolation method.
However it's writing a specific case when the general case seems to work fine which I thought was the case here. With integer indices, `gamma` is zero thus the interpolation would gives the left element.

But maybe it was made to handle some cases such as `uint8`.
Especially, I didn't pay that much attention to lower, higher, nearest which should preserve dtype.

Should I re-add this Integer handling ?

seberg replied:
Please do, I think it makes more sense for the discrete methods in general to be honest.

seberg commented on numpy/lib/function_base.py:
```diff
@@ -52,6 +53,187 @@
+        lambda gamma, _: _get_gamma_mask(shape=gamma.shape,
+                                         default_value=1.,
+                                         value=0.,
+                                         where=gamma == 0),
```
Should this be implemented as a correction to the first formula?  So that `gamma` can just be zeros (or even a special value `None` for non-interpolating variations?

seberg replied:
The same goes for `closest_observation`, which is also a discrete method (without any interpolation).

bzah replied:
I think it can be part of the first formula, but it is again to match what is described by H&F paper and ease understanding of this code.
Is there a benefit to integrate it in the first formula ?

seberg replied:
The reason I suggested it was as a way to skip the linear interpolation step for the discrete methods.  Unless that is the easiest thing to achieve that, there is no point.

bzah replied:
it's done.

seberg reviewed (commented):
Some smaller comments mainly, the only real (and harmless) issue I found the `nan` return path returning double precision always.  And some questions in case you have an opinion/insight.

The API discussion/issues I am still concerned about:
* I think we tended towards not making the Enum public.  We could keep it private, or just make it a private dictionary (e.g. of tuples) also simplify the lookup.
* I am not sure that `inclusive` is a reasonable name :(.  But maybe that can also be a followup.
* As noted in-line, I am unsure that constant outlier values are the best approach?  If we have doubt, an exception would be a safer choice?
* We should probably try to restore the conservation of the dtype for the discrete (the legacy `nearest`, `lower` and `upper` methods, as well as inverted-cdf and closest-observation)

In general I am very happy with the changes!  It is a big refactor and step in the right direction and I don't want to nitpick details too much :/.  I am can help out with the next steps, especially those related to the API, but input on some of the inline questions would be great.

(I would like to fix up the docs a bit more, but I think we are better off pushing that into a follow up as well.)

bzah commented on numpy/lib/function_base.py:
```diff
@@ -52,6 +53,187 @@
+        lambda n, quantiles: (n * quantiles) - 1,
+        lambda gamma, _: _get_gamma_mask(shape=gamma.shape,
+                                         default_value=1.,
+                                         value=0.5,
```
Fuzz has been removed (see https://github.com/numpy/numpy/pull/19857#issuecomment-919808623), I had a discussion with some people from R mailing list and we couldn't really understand the utility of it. 
I can send you by emails our discussion if you are interested.

bzah commented on numpy/lib/function_base.py:
```diff
@@ -4014,122 +4192,212 @@ def _quantile_is_valid(q):
+   """
+    return array_size * quantiles + (
+            alpha + quantiles * (1 - alpha - beta)
+    ) - 1
```
Ok for `array_size` -> `n`

The `-1` should be outside the bracket. It is used to correct the index which otherwise would be for an array starting at 1 (like in R)

I personally like exposing the whole formula because it looks like the formula in the original paper, which IMO help understand what's going on.
I had to rewrite multiple times `_lerp` (now `_linear_interpolation_formula`) so I think I understand what you mean but I don't know if factorizing `quantile` will mess with the boundary.

bzah commented on numpy/lib/function_base.py:
```diff
@@ -4014,122 +4192,212 @@ def _quantile_is_valid(q):
+    next_indexes = np.asanyarray(previous_indexes + 1)
+    indexes_above_bounds = virtual_indexes >= valid_values_count - 1
+    # When indexes is above max index, take the max value of the array
+    if indexes_above_bounds.any():
+        previous_indexes[indexes_above_bounds] = -1
+        next_indexes[indexes_above_bounds] = -1
+    # When indexes is below min index, take the min value of the array
+    indexes_below_bounds = virtual_indexes < 0
+    if indexes_below_bounds.any():
+        previous_indexes[indexes_below_bounds] = 0
+        next_indexes[indexes_below_bounds] = 0
```
Firstly, it is now necessary to handle such cases because the new virtual_index formula may produce such results. This was impossible before.

Now, I think extrapolation is a bad idea because to the user would probably not expect values outside of the sample min max boundaries when doing an interpolation.

I don't like throwing an error either because depending on the interpolation method chosen, it would sometimes throw sometimes not throw the error. For example method 7 would never throw the error (which makes it a very valid default value method).
But when the user get the error, it will be hard for him/her to understand what to do to avoid this.

I admit my solution is a bit home brewed but for what is worth, 
- I believe it give a similar behavior of what R does (not 100% sure though).
- scipy::mquantiles seems to have the same behavior (see the use of `clip` https://github.com/scipy/scipy/blob/v1.7.1/scipy/stats/mstats_basic.py#L2659-L2784)

bzah commented:
@seberg 
In H&F paper, instead of "inclusive" it is traced to Gumble (1939). 
We are already using authors names for methods 5 and 6, so method 7 could be method Gumble I suppose.

seberg commented:
Yeah `gumble` is a decent name.  Right now (keeping `interpolation` intact), we can also just have only `linear` (it is the default anyways, so the name is probably only relevant because it is commonly read in the function signature).  I am also wondering about `sample` since the method has no correction to achieve an unbiased population estimate.

I am OK with the clipping behaviour, especially since I guess that is what R does. Just a bit unsure if there might be good reasons to pick the strict version.

Anyway, both are things that seem fine and we can do an executive decision on; although I will pass by the mailing list once at some point, just in case.

bzah commented on numpy/lib/function_base.py:
```diff
@@ -3762,20 +3943,75 @@ def percentile(a, q, axis=None, out=None,
         `a` after this function completes is undefined.
 
-    interpolation : {'linear', 'lower', 'higher', 'midpoint', 'nearest'}
+    interpolation : { 'inverted_cdf', 'averaged_inverted_cdf',
```
Is it too verbose ?

charris commented:
Wrong indents in the docstrings.

bzah commented:
@charris Yeah I saw that, but I could not find how to run the doctrings linter locally, so my "Fix issue with doc" did not fix anything.
I will try to work on this soon.

charris commented:
My guess is that the docstring parser cannot handle the multiline options for `interpolate`. Maybe just type it as `str` and list the options below, naming the default.

charris committed:
DOC: fix docstrings.

Hopefully fix the docstrings of percentile, nanpercentile,
quantile, and nanquantile so that CircleCI passes.

abel-bzz committed:
Fix _lerp

- some changes were unrelated to the PR and have been reverted, including, renaming and moving the logic around.
- Also renamed _quantile_ureduce_func to its original name

abel-bzz committed:
DOC: Improve quantile documentation

Also removed unused imports

seberg commented:
> I still don't know if we should and how to avoid the 4 replications of quantile doc.

I guess, we could do two things (not sure what we should):
* We could put the docs into a string and use string formatting to include it in the 4 places (just to reduce the copy-pasting).
* I personally think it it is probably OK to refer to `np.quantile` for long the explanation that is currently in the `Notes`.  Especially the `nan` versions of the function very often refer to the non-nan ones anyway.  I.e. we could also duplicate it for `percentile` and `quantile`, but not in the NaN versions if that seems better.

charris commented:
@seberg Two looks like a reasonable solution, but I didn't intend to delay this PR to deal with the documentation repeats :) This isn't the only place it occurs.

charris commented:
The circleCI problem is because of the SVML submodule, the other failing tests look like an example failure on a return value not being converted to a scalar.

abel-bzz committed:
Fix issue with nan scalar

Also added unit test for it.

seberg reviewed (commented):
Thanks!  Since there was no more discussion on the mailing list, I am in favor of pushing forward soon.  A bunch of nits (take them or ignore them!).

What I would like is the test cleanup and additional tests for the "extrapolation" behaviour mainly.  I am also slightly worried about bugs with N-D inputs because that one `partition` call seems to miss the `axis` and the tests should have noticed that (if true)?
It might make sense to add another test, I suppose one discrete and one non-discrete method would be sufficient.

In any case, the code looks good and further cleanups (if desired) can be done as followups IMO.  I am tempted to check code coverage, but I doubt we have uncovered code and even if, its probably not the interesting/tricky stuff...

seberg commented:
@bzah thanks, but I think we are pretty much settling anyway.  So I would prefer to merge it soon, so we can comfortable make the next release and maybe consider renaming the kwarg before it (also gives some time to flush out bugs).
The only thing I am personally worried about is the "extrapolation" part, mainly because I don't  have an opinion and just noticed Python actually extrapolates:
```
In [5]: statistics.quantiles([15.0, 20.0, 35.0, 40.0, 50.0], n=10)
Out[5]: [13.0, 16.0, 19.0, 26.0, 35.0, 38.0, 42.0, 48.0, 54.0]  # First value smaller 15.
```

Otherwise, no biggy, but chances are this will miss the next release (which will delay it by about half a year).

abel-bzz committed:
MAINT: Clean following PR comments

abel-bzz committed:
TST: Add extrapolation tests

BvB93 reviewed (commented):
I'm not sure if this will require a rebase, bit could you add the new interpolation kinds to the stub file?

https://github.com/numpy/numpy/blob/ba3664766d1f8c6b2521b82edd717a2f2be974f2/numpy/lib/function_base.pyi#L503-L509

abel-bzz committed:
TST: Add parametrize for interpolation methods

abel-bzz committed:
DOC: Update _InterpolationKind

charris commented:
> 1. Lint works on the Ci but locally there was a few problems in `numpy/array_api`.
Don't worry about it.

> 2. Tests were workings locally but on the github CI `tests/test_nditer.py:3151:` fails.
That test is flaky.

abel-bzz committed:
TST: Make use of clongdouble and longdouble

On some platforms float128 and complex256 do not exist.
Using (c)longdouble aliases should work on all platforms.

seberg committed:
MAINT: Remove unnecessary import of _QuantileInterpolation

seberg commented:
Lint is just 1 character to long and slightly weird indentation.  I tried to reorganize the docs to use "templates"... happy to undo or modify it, but while not perfect it seemed better than repeating those long strings 4 times.

Otherwise, there is no change (just a rebase).  I feel this was ready, the main small annoyance is that I still think we may want to rename the keyword argument to `method`, since "interpolation" is techncially not even correct (interpolation forgets the fact that the population extend is larger than the sample extend).

So the only squirms I have is: If I put it in now, should I make sure we do that rename before the 1.22 release?

seberg committed:
DOC: Refer to the quantile/percentile notes for nan versions

seberg commented:
Nevermind, using f-strings doesn't work for docstrings (you would have to use the trick to replace `function.__doc__`.  So just removed the Notes from the nan versions instead with a link to the other. Planning to merge it today (pending fixup's in case the docs misrender).

seberg commented:
Thanks @bzah.  Lets put this in.  If there are any remaining concerns, please open an issue or mention them here so we can followup before the next release hopefully.

I am planning to follow up with a rename of `interpolation` to `method` before the NumPy 1.22 release (I think "interpolation" is not a good name and do not want to avoid users of the new methods to require moving.)

If anyone wants to take a closer look at the documentation to see if it can be improved to provide more guidelines that would be great.  (I think of example the default method could be considered a sample quantile/percentile, while most others are various population estimates, if that is correct it might be helpful to explain earlier/clearer?)

jbrockmendel commented:
@seberg are the quantile changes something we need to adapt to on our (pandas) end or to address here?

seberg commented:
@jbrockmendel I am honestly not quite sure yet.  The formula used for calculating seems "index" not ideal (potentially a few ULPs less imprecise than it could be for the default method).  I am not sure if we can reorder the formula in general, or whether that would lead to worse characteristics for some methods.

@bzah suggested using a different (simplified/old one) formula for the current default, I suppose we could go through all methods to see if we can clearly simplify a few more.  If the changes in pandas seem too strange to be true, I am happy to take that as enough reason that we should specialize at least the default.

One thing I am curious about: If this is the reason for the change, I wonder if the code that tests this will misbehave for all methods except the ones NumPy previously implemented!  For all other methods, the repeating samples would change the result (my working hypothesis is that this happens, but I am not familiar with `Series.expanding`).

jbrockmendel commented:
>  If the changes in pandas seem too strange to be true

The few test failures I've looked at have been small floating point errors https://github.com/pandas-dev/pandas/runs/4131499766?check_suite_focus=true#step:7:104 e.g. 

```
>           assert q1 == q2
E           assert 0.18000000000000005 == 0.18000000000000002
```

seberg commented:
@bzah I just realized that we already need the `virtual_indices` formula in the dictionary anyway for the non-interpolating methods.  So I think the idea of special casing the formula for `"linear"` is good (maybe with a very short comment of why?)

Is `alpha == beta == 1` special or is `alpha == beta` maybe already enough to have a simplified version?

Do you want to make a quick PR?

bzah commented:
Yup, I'm on it.
